src/main/java/org/wildfly/security/credential/source/CredentialStoreCredentialSource.java

```diff
@@ -64,14 +94,17 @@ public SupportLevel getCredentialAcquireSupport(final Class<? extends Credential
 
     public <C extends Credential> C getCredential(final Class<C> credentialType, final String algorithmName, final AlgorithmParameterSpec parameterSpec) throws IOException {
         final C credential;
+        final Class<C> crType;
         try {
-            credential = credentialStore.retrieve(alias, credentialType);
+            createAndInitCredentialStore();
+            crType = credentialType != null ? credentialType : (Class<C>) credentialStoreReference.getType();
```

Comment:
I don't quite understand the credential store reference, which is something we should talk about, however I will say that in this particular case there are a couple of problems... First, credentialType can not be null for CredentialSource.  Second, you can't cast Class objects like this safely; instead you would do something like this: ```credentialStoreReference.getType().asSubclass(credentialType);``` which will yield a ```Class<? extends C>``` which would be good enough for this particular case.